controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,25 +152,229 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
-				machine.Namespace = r.ReconciliationSubject.Namespace
-				if err := r.K8sClient.Delete(r.RequestCtx, machine); err != nil {
+			if ref.Kind == kindEtcdadmCluster {
+				etcdMachineFound = true
+				// get etcdadmcluster by using machine's etcdadmcluster ownerReference
+				etcdadmCluster := &unstructured.Unstructured{}
+				etcdadmCluster.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: ref.Name}, etcdadmCluster); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				csMachineTemplateName, err := getCsMachineTemplateName(machine, etcdadmCluster, ref.Name)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				newCsMachineTemplateName, err := cloneCsMachineTemplateWithNewName(r, machine.Namespace, csMachineTemplateName, ref.Name)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// patch etcdadmcluster to refer newly created cloudstack machine template, which will trigger an etcdadmcluster rolling upgrade.
+				if _, err := controllerutil.CreateOrPatch(r.RequestCtx, r.K8sClient, etcdadmCluster,
+					func() error {
+						return unstructured.SetNestedField(
+							etcdadmCluster.Object,
+							newCsMachineTemplateName,
+							"spec", "infrastructureTemplate", "name",
+						)
+					}); err != nil {
 					return ctrl.Result{}, err
 				}
+				r.Log.Info(fmt.Sprintf("failuredomain delete: etcdadmcluster infrastructureTemplate cloudstackmachinetemplate name %s patched", newCsMachineTemplateName))
+				break
 			}
 		}
+		if etcdMachineFound {
+			break
+		}
 	}
-	if len(machines.Items) > 0 {
-		return r.RequeueWithMessage("FailureDomain still has machine(s) in it.")
+
+	return ctrl.Result{}, nil
+}
+func triggerMachineDeploymentRollout(machines []infrav1.CloudStackMachine, r *CloudStackFailureDomainReconciliationRunner) (ctrl.Result, error) {
+	for _, machine := range machines {
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == kindMachineSet {
+
+				// get machine deployment by using cloudstack machine's label info
+				md := &clusterv1.MachineDeployment{}
+				mdName, ok := machine.Labels[clusterv1.MachineDeploymentLabelName]
+				if !ok {
+					return ctrl.Result{}, errors.Errorf("cloudstack machine %s misses label: %s", machine.Name, clusterv1.MachineDeploymentLabelName)
+				}
+
+				if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: machine.Namespace, Name: mdName}, md); err != nil {
+					return ctrl.Result{}, err
+				}
+
+				// add an annotation restartedAt in machine deployment if such one not already added
+				// this will trigger an immediate machine deployment rollout
+				patcher, err := patch.NewHelper(md, r.K8sClient)
+				if md.Spec.Template.Annotations == nil {
+					md.Spec.Template.Annotations = map[string]string{}
+				}
+				timeNowStr := time.Now().Add(time.Second * time.Duration(10)).Format(time.RFC3339)
+				if err != nil {
```

Comment:
I think error check blocks should be immediately after the statement that sets the error variable.